knt-backend-go/knt/internal/knt_database/database.go

```diff
@@ -0,0 +1,47 @@
+package kntdatabase
+
+import (
+	"database/sql"
+	"log"
+)
+
+func GetAllProducts(db *sql.DB) ProductList {
+	var l ProductList
+	l.Products = genericQuery(db, "select * from product", func(r *sql.Rows) Product {
+		var p Product
+		err := r.Scan(&p.Id, &p.Price, &p.Name)
+		if err != nil {
+			log.Fatal(err)
+		}
+		return p
+	})
+	return l
+}
+
+func GetAllUsers(db *sql.DB) UserList {
+	var l UserList
+	l.Users = genericQuery(db, "select * from user", func(r *sql.Rows) User {
+		var p User
+		err := r.Scan(&p.Id, &p.FirstName, &p.LastName, &p.VunetId, &p.Password, &p.Balance)
+		if err != nil {
+			log.Fatal(err)
+		}
+		return p
+	})
+	return l
+}
+
+//Generic query handles scaling through all the rows. Functions using this query have to provide
+//Code that will handle each row. Appending and everything else is handles by the function
+func genericQuery[K any](db *sql.DB, queryString string, f func(*sql.Rows) K) (output []K) {
+	rows, err := db.Query(queryString)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer rows.Close()
+	var slice []K
+	for rows.Next() {
+		slice = append(slice, f(rows))
+	}
+	return slice
+}
```

Comment:
This is somewhat cursed, but as long as it is **_used only_** inside of the `database.go` file, as to avoid code duplication, it is fine.

I think adding functions that handle each case (e.g. consuming product X, quantity Y, user Z) would be good to keep overview on the code.

Reply:
Yes the function is not exported so it can only be used there.

That is the plan for the future revisions, for now I focused on only implementing the functionality that was implemented in the previous backend